nebula/ui/components/VPNFlickable.qml

```diff
@@ -158,4 +87,90 @@ Flickable {
         }
     }
 
+    function ensureVisible(item) {
+        if(ensureVisAnimation.running) {
+            ensureVisAnimation.stop()
+        }
+
+        let yPosition = item.mapToItem(contentItem, 0, 0).y
+        if (!contentExceedsHeight || item.skipEnsureVisible || yPosition < 0) {
+            return
+        }
+
+        const isItemBehindNavbar = item.mapToItem(window.contentItem, 0, 0).y + item.height > window.height - VPNTheme.theme.navBarHeightWithMargins
+        const buffer = navbar.visible && isItemBehindNavbar ? VPNTheme.theme.navBarHeightWithMargins : 20
+        const itemHeight = Math.max(item.height, VPNTheme.theme.rowHeight) + buffer
+        let destinationY
+
+        //When focusing on an item behind the navbar in the downward direction, make sure it is not blocked by the navbar
+        if(navbar.visible && isItemBehindNavbar) {
+            const distanceToGetOutFromBehindNavbar = item.mapToItem(window.contentItem, 0, 0).y + item.height - (window.height - buffer)
+
+            destinationY = contentY + distanceToGetOutFromBehindNavbar
+        }
+        else {
+            //When focusing on an item that is off screen in the upward direction
+            if (yPosition < vpnFlickable.contentY) {
+                destinationY = yPosition - buffer
+            }
+        }
+
+        if (typeof(destinationY) === "undefined") return;
+
+        //For some reaason during tutorials, mobile devices won't perform an animated scroll??
+        if(VPNTutorial.playing && window.fullscreenRequired()) {
+            contentY = destinationY
+        }
+        else {
+            ensureVisAnimation.to = destinationY
+            ensureVisAnimation.start()
+        }
+
+        return
+    }
+
+    function recalculateContentHeight() {
+        //Absolute y coordinate position of the scroll view
+        const absoluteYPosition = mapToItem(window.contentItem, 0, 0).y
+        //Portion of the screen that a view's contents can reside without interfering with the navbar
+        const contentSpace = window.height - VPNTheme.theme.navBarHeightWithMargins
+
+        //Checks if the flickable AND it's content interferes with the navbar area (bottom 146px for iOS, bottom 128px for other platforms)
+        //If it does, we pad the flickable's contentHeight with whatever bottom padding is needed so that there is always 48px (aka theme.navBarTopMargin)
+        //between the bottom of the flickable's content and the navbar
+        if (navbar.visible && absoluteYPosition + height >= contentSpace
+                && flickContentHeight + absoluteYPosition >= contentSpace) {
+            vpnFlickable.contentHeight = flickContentHeight + (flickContentHeight >= height ? VPNTheme.theme.navBarHeightWithMargins : (absoluteYPosition + flickContentHeight) - contentSpace + (height - flickContentHeight))
```

Comment:
Suggestion (non-blocking): This line is really hard to grasp. Could these calculations could be broken down and assigned to descriptive variables?

Reply:
Yeaaaa I sort of expected this suggestion 😅 Naming is hard but I'll try to brainstorm some variable names, and in the meantime, there are comments for some of the less self-descriptive calculations.